### PR TITLE
fix: Add `--no-schema` option to `litestar routes` command.

### LINKED ISSUE #1614: Enhancement: Adding `--no-schema` option to `litestar routes` command
### Summary

Add `--no-schema` option to `litestar routes` to hide schema routes.

### Basic Example

`litestar routes` currently display all routes information including schemata.

```shell
> litestar routes

Using Litestar app from app:app
/ (HTTP)
├── heartbeat async GET
└── options_handler sync OPTIONS
/detect (HTTP)
├── detect async POST
└── options_handler sync OPTIONS
/detect-without-roi (HTTP)
├── detect_without_roi async POST
└── options_handler sync OPTIONS
/schema (HTTP)
├── root sync GET
└── options_handler sync OPTIONS
/schema/elements (HTTP)
├── stoplight_elements sync GET
└── options_handler sync OPTIONS
/schema/openapi.json (HTTP)
├── retrieve_schema_json sync GET
└── options_handler sync OPTIONS
/schema/openapi.yaml (HTTP)
├── retrieve_schema_yaml sync GET
└── options_handler sync OPTIONS
/schema/redoc (HTTP)
├── redoc sync GET
└── options_handler sync OPTIONS
/schema/swagger (HTTP)
├── swagger_ui sync GET
└── options_handler sync OPTIONS
```

Sometimes, I just want to glance at routes only **I** built, not including schema routes Litestar automatically generated.

Adding an option such as `--no-schema` to `litestar routes` would be appreciated.

```shell
> litestar routes --no-schema

Using Litestar app from app:app
/ (HTTP)
├── heartbeat async GET
└── options_handler sync OPTIONS
/detect (HTTP)
├── detect async POST
└── options_handler sync OPTIONS
/detect-without-roi (HTTP)
├── detect_without_roi async POST
└── options_handler sync OPTIONS
```

I am happy to implement this within a day or two if the idea gets no objection 😀

### Drawbacks and Impact

Since it is an option, it will not bring any backward incompatible behaviors.

### Unresolved questions

_No response_

<a href="https://polar.sh/litestar-org/litestar/issues/1614">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/1614/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by solzard**
### Pull Request Checklist

[//]: # "Please review the [Litestar contribution guidelines](https://github.com/litestar-org/litestar/blob/main/CONTRIBUTING.rst) for this repository."

- [x] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

### Description

[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

This PR adds --no-schema option to litestar routes CLI command to exclude schema routes from the list.

### Close Issue(s)

[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

#1614

### CODE REVIEW

**Comment by peterschutt** [comment]
Maybe this is too rigid - we document how to rename the path the openapi docs are served at here: https://docs.litestar.dev/2/usage/openapi.html#subclassing-openapicontroller - and so this wouldn't work if that was done.

How about something more general, like `--exclude=/schema/.*` where the value is interpreted as a regex and any route that matches is excluded from output?

@litestar-org/maintainers, thoughts?

**Comment by solzard** [comment]
How about this modification honoring `app.openapi_config`?

```python
from litestar.app import DEFAULT_OPENAPI_CONFIG

openapi_config = app.openapi_config if app.openapi_config is not None else DEFAULT_OPENAPI_CONFIG
schema_path = openapi_config.openapi_controller.path  # /schema or user-defined path

for route in sorted(app.routes, key=lambda r: r.path):
    if no_schema and route.path.startswith(schema_path):
        continue

    ...
```

File: litestar/cli/commands/core.py [hunk 1]
```diff
@@ -98,12 +98,16 @@ def run_command(
 
 
 @command(name="routes")
-def routes_command(app: Litestar) -> None:  # pragma: no cover
+@option("--no-schema", help="Exclude shema routes", is_flag=True, default=False)
```
Comment (JacobCoffee) [inline litestar/cli/commands/core.py:101]:
```suggestion
@option("--no-schema", help="Exclude schema routes", is_flag=True, default=False)
```

**Review by JacobCoffee — CHANGES_REQUESTED** [review]

**Comment by peterschutt** [comment]
> Or, how about this modification honoring `app.openapi_config`?
> 
> ```python
> ...
> 
> from litestar.app import DEFAULT_OPENAPI_CONFIG
> 
> openapi_config = app.openapi_config or DEFAULT_OPENAPI_CONFIG
> schema_path = openapi_config.openapi_controller.path  # /schema or user-defined path
> 
> for route in sorted(app.routes, key=lambda r: r.path):
>     if no_schema and route.path.startswith(schema_path):
>         continue
> 
>     ...
> ```

I still prefer the regex based exclude solution as I think it is generally more useful, however I haven't had much to do with developing this CLI, so others in here would have an opinion that would outweigh mine.

**Comment by provinzkraut** [comment]
> however I haven't had much to do with developing this CLI, so others in here would have an opinion that would outweigh mine.

I like the regex idea, and I want to throw another one into the ring: 
Make `--no-schema` the default, and instead add a switch to include those routes. The regex thing could then be added independently.

**Comment by peterschutt** [comment]
> I like the regex idea, and I want to throw another one into the ring:
> Make `--no-schema` the default, and instead add a switch to include those routes. The regex thing could then be added independently.

I think it would make sense to do both at the same time then. Implement the regex exclude option, and add a default regex exclude that excludes the configured schema route path, and have that behavior configurable with `--no-schema=False`.

Will you be continuing with this @solzard? Absolutely no pressure, just curious.

**Comment by solzard** [comment]
Yes, I'm planning to do it this weekend.
Two questions:
1. If `--exclude` option is passed, should it override the default schema regex? Or, merge with it?
2. Should the CLI accepts several `--exclude` options? (e.g. `litestar routes --exclude=/foo/* --exclude=/bar/*`)

**Comment by peterschutt** [comment]
> Yes, I'm planning to do it this weekend. Two questions:
> 
>     1. If `--exclude` option is passed, should it override the default schema regex? Or, merge with it?
> 
>     2. Should the CLI accepts several `--exclude` options? (e.g. `litestar routes --exclude=/foo/* --exclude=/bar/*`)

1. It should merge with it. Given the default behavior will be to exclude schema routes, it would be surprising if adding an `--exclude` resulted in display of the schema routes.
2. Yep.

Thanks heaps!

**Comment by Goldziher** [comment]
@solzard ?

**Comment by provinzkraut** [comment]
Closing this as stale

**Pull request closed without merging**